enhancements/network/user-space-pod-iface-and-api-library.md

```diff
@@ -0,0 +1,262 @@
+- This could, secondarily, also be used to support user space data plane
+technologies like ovs-dpdk, but its not the same, and its not our primary goal.
+- Non-SR-IOV devices (vDPA and vhost-user interfaces) require additional
+information. However, these devices are not required for this enhancement.
+That being said, some of the upstream work this enhancement is pulling from does
+include those interfaces. So the API Library may include some of the additional
+work for these interfaces.
+
+## Proposal
+
+As described above, there are two parts to this enhancement. Work is need to
+pass additional data to the pod to enable user space applications to run
+properly. Work is also needed in the pod to ease the consumption of the data.
+Each item will be described below.
+
+### Passing Data to the Pod
+
+Currently, data is passed to the pod in several forms:
+- **PCI Addresses**: SR-IOV Device Plugin passes the list of PCI Addresses in
+the an environmental variable to the pod.
+- **Interface Data**: Multus populates the pod annotation with a
+'*k8s.v1.cni.cncf.io/networks-status*' field. This annotation contains the list
+of interfaces, and for each interface, the MAC address, list of IP addresses
+(not CIDR), interface (net1, net2, etc), name, etc. If the Kubernetes Downward
+API is enabled, this information is available to the pod.
+- **CPU List**: The dedicated CPU list can be retrieved by reading the
+'*cpuset*' from the cGroup in the pod.
+
+For user space applications running in pod, the following information is still
+needed:
+- **Hugepage Information**: The type and number of hugepages allocated to the
+pod.
+- **Additional Interface Data**: The '*k8s.v1.cni.cncf.io/networks-status*'
+contains a '*name*', which is the network in the CRD, but doesn't have enough
+information to determine which PCI Adress is associated with which interface.
+So addition information is needed o make that association within the pod.
+
+#### Hugepage Information
+
+There was an upstream Kubernetes PR,
+[86102](https://github.com/kubernetes/kubernetes/pull/86102),
+to enhance the Kubernetes Downward API to add hugepage information to the set
+of supported fields. This PR was closed due to inactivity. The plan is to
+resubmit the PR and try to push the functionality into Kubernetes. It doesn't
+appear that there were any major objections to the PR, just lack of activity.
+
+If this feature can make it into upstream Kubernetes, then the pod will have
+hugepage information needed by user space applications.
+
+#### Additional Interface Data
+
+Work is being done in the Network Plumbing Working Group (NPWG) to standardize
+how data for secondary interfaces is passed to a pod. NPWG is where the standard
+is defined that Multus is implemented too. See:
+[Kubernetes Network Custom Resource Definition De-facto Standard](https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit?usp=sharing).
```

Comment:
should probably reference something in https://github.com/k8snetworkplumbingwg/multi-net-spec now?

Reply:
Better. Done.